libminifi/src/utils/file/FileMatcher.cpp

```diff
@@ -0,0 +1,273 @@
+    ++it;
+  }
+}
+
+static bool matchGlob(std::string::const_iterator pattern_it, std::string::const_iterator pattern_end, std::string::const_iterator value_it, std::string::const_iterator value_end) {
+  // match * and ?
+  for (; pattern_it != pattern_end; ++pattern_it) {
+    if (*pattern_it == '*') {
+      do {
+        if (matchGlob(std::next(pattern_it), pattern_end, value_it, value_end)) {
+          return true;
+        }
+      } while (advance_if_not_equal(value_it, value_end));
+      return false;
+    }
+    if (value_it == value_end) {
+      return false;
+    }
+    if (*pattern_it != '?' && *pattern_it != *value_it) {
+      return false;
+    }
+    ++value_it;
+  }
+  return value_it == value_end;
+}
+
+FileMatcher::FilePattern::DirMatchResult FileMatcher::FilePattern::matchDirectory(DirIt pattern_it, DirIt pattern_end, DirIt value_it, DirIt value_end) {
+  for (; pattern_it != pattern_end; ++pattern_it) {
+    if (is_this_dir(*pattern_it)) {
+      continue;
+    }
+    if (*pattern_it == "**") {
+      if (std::next(pattern_it) == pattern_end) {
+        return DirMatchResult::TREE;
+      }
+      bool matched_parent = false;
+      // any number of nested directories
+      do {
+        skip_if(value_it, value_end, is_this_dir);
+        auto result = matchDirectory(std::next(pattern_it), pattern_end, value_it, value_end);
+        if (result == DirMatchResult::TREE || result == DirMatchResult::EXACT) {
+          return result;
+        }
+        if (result == DirMatchResult::PARENT) {
+          // even though we have a parent match, there may be a "better" (exact, tree) match
+          matched_parent = true;
+        }
+      } while (advance_if_not_equal(value_it, value_end));
+      if (matched_parent) {
+        return DirMatchResult::PARENT;
+      }
+      return DirMatchResult::NONE;
+    }
+    skip_if(value_it, value_end, is_this_dir);
+    if (value_it == value_end) {
+      // we used up all the value segments but there are still pattern segments
+      return DirMatchResult::PARENT;
+    }
+    if (!matchGlob(pattern_it->begin(), pattern_it->end(), value_it->begin(), value_it->end())) {
+      return DirMatchResult::NONE;
+    }
+    ++value_it;
+  }
+  skip_if(value_it, value_end, is_this_dir);
+  if (value_it == value_end) {
+    // used up all pattern and value segments
+    return DirMatchResult::EXACT;
+  } else {
+    // used up all pattern segments but we still have value segments
+    return DirMatchResult::NONE;
+  }
+}
+
+bool FileMatcher::FilePattern::match(const std::string& directory, const optional<std::string>& filename) const {
+  auto value = split(directory, {"/", "\\"});
+  auto result = matchDirectory(directory_segments_.begin(), directory_segments_.end(), value.begin(), value.end());
+  if (!filename) {
+    if (excluding_) {
+      if (result == DirMatchResult::TREE && file_pattern_ == "*") {
+        // all files are excluded in this directory
+        return true;
```

Comment:
I did some refactoring, not sure if it made things less complicated

Reply:
Yes, the new version is clearer.  However, I would completely separate the directory matching from the file matching; there will be some duplication of code, but I think it is going to be easier to understand and maintain.  This is what I have in mind: https://github.com/fgerlits/nifi-minifi-cpp/commit/c271b0f612b704bcb72e48afb709eaa4b8f3a377

Reply:
I experimented with separating the two, but to my eyes it was little more cumbersome (as we have to iterate twice) but it could make things easier to read, cherry-picked your commit